### PR TITLE
feat(metrics): show actor/table id info in grafana dashboard

### LINKED ISSUE #8814: Expost actor id, fragment id, table id, table name mapping in grafana
Exposing the mapping among (fragment_id, actor_id, table_id, table_name) in Grafana via [Table panel](https://grafana.com/docs/grafana/latest/panels-visualizations/visualizations/table/) is beneficial for the following reasons:
- Most of our metrics are only tagged by some kind of id
- We oftenly want to know the executor type or the table name when looking at the metric during debugging
- These information is not available if the cluster is down but they are super helpful to debug an issue using historical metrics stored in Prometheus via Grafana.

Note that the storage cost of this mapping in Prometheus can be small if we only emit one data point on actor creation.

### PR DESCRIPTION
**Pull request by hzxa21**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR adds a "Actor/Table Id Info" panel in grafana dashboard to display the relationship among actor id, fragment id, table id and table name.

Implementation detail:
We will emit one dummy gauge value labelled by (actor id, fragment id, compute node address) and one dummay gauge value labelled by (table id, actor id, table name) by scaning all fragments every 60s via fragment manager in the meta node.

These labels will then be used by the grafana dashboard to plot a table to display the mapping info.

close #8814

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
~~- [ ] I have added necessary unit tests and integration tests~~
~~- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).~~
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [x] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Comment by hzxa21** [comment]
Example:
image

File: src/stream/src/from_proto/hash_agg.rs [hunk 1]
```diff
@@ -85,12 +85,21 @@ impl ExecutorBuilder for HashAggExecutorBuilder {
             vnodes.clone(),
         )
         .await;
-        let result_table = StateTable::from_table_catalog(
-            node.get_result_table().unwrap(),
-            store.clone(),
-            vnodes.clone(),
-        )
-        .await;
+        let table_catalog = node.get_result_table().unwrap();
```
Comment (chenzl25) [inline src/stream/src/from_proto/hash_agg.rs:88]: We miss the state storage here from L82 to L87.

**Review by chenzl25 — APPROVED** [review]
Rest LGTM!

**Review by BugenZhao — COMMENTED** [review]
This implementation looks error-prone and not that clear to me. 🤔 

Actually, all of the info recorded in the `from_proto` is known in the meta service, and this report can be simply generated by periodically querying the `FragmentManager`. There seems no need to be incrementally maintained on the compute node at all.

**Review by st1page — COMMENTED** [review]
agree with @BugenZhao some other concerns:
- in meta, we already have a unified visitor for each state table which is easier to maintain  https://github.com/singularity-data/risingwave/blob/main/src/meta/src/stream/stream_graph/visit.rs
- it is harder to maintain the metrics with the actor's lifetime in the compute node. I do not see the logic of the unregister when the actor is dropped or migrated to other nodes. even worse, just considering if the compute node is crushed, how to unregister all the state tables?

**Comment by st1page** [comment]
> agree with @BugenZhao some other concerns:
> 
> * in meta, we already have a unified visitor for each state table which is easier to maintain https://github.com/singularity-data/risingwave/blob/main/src/meta/src/stream/stream_graph/visit.rs and do not easy to miss some table.
> * it is harder to maintain the metrics with the actor's lifetime in the compute node. I do not see the logic of the unregister when the actor is dropped or migrated to other nodes. even worse, just considering if the compute node is crushed, how to unregister all the state tables?

**Comment by hzxa21** [comment]
> This implementation looks error-prone and not that clear to me. 🤔
> 
> Actually, all of the info recorded in the `from_proto` is known in the meta service, and this report can be simply generated by periodically querying the `FragmentManager`. There seems no need to be incrementally maintained on the compute node at all.

I agree that if RisingWave cluster is up we can easily get these info. However, I think it is important to get this info even when RisingWave cluster is down. For example, after user deletes the cluster or there are bugs causing the whole cluster to fail. From my recent debugging experience, not knowing the id mapping when the cluster is down makes debugging way harder since it is not easy to identify the correct metrics.

Mataining a similar collector in meta node to emit these info also works but I still think it is better to do it incrementally since all we need is the labels and the metric values are useless so there is no need to emit the same metrics over and over again. Also, maybe it is better to tag the actor id with CN address. If that is clearer to you, I can switch to this approach instead.

**Comment by st1page** [comment]
> However, I think it is important to get this info even when RisingWave cluster is down. For example, after user deletes the cluster or there are bugs causing the whole cluster to fail. From my recent debugging experience, not knowing the id mapping when the cluster is down makes debugging way harder since it is not easy to identify the correct metrics.

Maybe log is a better choice for this case? RisingWave indeed lack logs in any level :hot_face:

**Comment by hzxa21** [comment]
> in meta, we already have a unified visitor for each state table which is easier to maintain https://github.com/singularity-data/risingwave/blob/main/src/meta/src/stream/stream_graph/visit.rs

I am not aware of that. This is great! I guess I will emit metrics in meta during DDL and scale then.

> it is harder to maintain the metrics with the actor's lifetime in the compute node. I do not see the logic of the unregister when the actor is dropped or migrated to other nodes. even worse, just considering if the compute node is crushed, how to unregister all the state tables?

This is not an issue when implementing a collector to emit incremental metric values instead of defining metrics explicitly. The reasons are: metrics unregisteration is needed only when we have maintain some in-memory states in the node (via defining the metrics explicitly) since all it does is to clear memory, not clear the historical metric values in the metrics storage backend.

This is the exact reason why I think implementing a collector to emit the metric incrementally is preferred.

**Comment by hzxa21** [comment]
> > However, I think it is important to get this info even when RisingWave cluster is down. For example, after user deletes the cluster or there are bugs causing the whole cluster to fail. From my recent debugging experience, not knowing the id mapping when the cluster is down makes debugging way harder since it is not easy to identify the correct metrics.
> 
> Maybe log is a better choice for this case? RisingWave indeed lack logs in any level 🥵

+1 for adding more logs. But I still think it is better to visualize the mapping somewhere before we have implemented a better alternative of logging infra instead of using CloudWatch... I don't like searching for millions of logs via CloudWatch just to get the actor info :(

**Comment by BugenZhao** [comment]
> I agree that if RisingWave cluster is up we can easily get these info. However, I think it is important to get this info even when RisingWave cluster is down.

Agree. But by **periodically** querying the `FragmentManager` and then reporting to Prometheus on the meta service, I think we can also achieve this goal?

> Mataining a similar collector in meta node to emit these info also works but I still think it is better to do it incrementally since all we need is the labels and the metric values are useless so there is no need to emit the same metrics over and over again.

Since this is infrequent and lightweight in normal cases, I think full scan can be more friendly to programming. 🤔 We don't need to worry about missing some config-change paths and the result should always be accurate, as we have already put much effort on maintaining the consistency between meta store and physical graph on the compute nodes.

> Also, maybe it is better to tag the actor id with CN address.

Yes. We can do this on the meta service by `ClusterManager` as well.

> he reasons are: metrics unregisteration is needed only when we have maintain some in-memory states in the node (via defining the metrics explicitly) since all it does is to clear memory, not clear the historical metric values in the metrics storage backend.

I guess this may not be the only reason. 🤔 The lifetime of each actor should still be important to us. Imagine that there was a lot of scaling happened to a fragment, and then the dashboard will show us a lot of actor ids, most of which are already dead and noisy to us.

**Comment by hzxa21** [comment]
Modified the implementation to list table fragments every 60s to report full metrics in a background task in meta node.
@BugenZhao @st1page PTAL.

**Review by BugenZhao — APPROVED** [review]
LGTM!

**Review by st1page — APPROVED** [review]
LGTM, and maybe we can store definition of the query in the tableFragment to get more information? c.c. @BugenZhao @yezizp2012

**Review by yezizp2012 — APPROVED** [review]

**Comment by yezizp2012** [comment]
> LGTM, and maybe we can store definition of the query in the tableFragment to get more information? c.c. @BugenZhao @yezizp2012

LGTM, it's useful when cluster is down. We can simply get the definition from `CatalogManager`, in the same way that we get the cluster infos in this PR.

**Comment by BugenZhao** [comment]
> LGTM, and maybe we can store definition of the query in the tableFragment to get more information? c.c. @BugenZhao

Yes. Currently we have that on each `Actor`. 😄

https://github.com/risingwavelabs/risingwave/blob/d990fabb651e0a4979566fe83f0606fa92ebb28c/proto/stream_plan.proto#L654-L655

**Comment by st1page** [comment]
> > LGTM, and maybe we can store definition of the query in the tableFragment to get more information? c.c. @BugenZhao
> 
> Yes. Currently we have that on each `Actor`. smile
> 
> https://github.com/risingwavelabs/risingwave/blob/d990fabb651e0a4979566fe83f0606fa92ebb28c/proto/stream_plan.proto#L654-L655

Great! is it also work for `create sink` statement?

**Comment by yezizp2012** [comment]
> > > LGTM, and maybe we can store definition of the query in the tableFragment to get more information? c.c. @BugenZhao
> > 
> > 
> > Yes. Currently we have that on each `Actor`. smile
> > https://github.com/risingwavelabs/risingwave/blob/d990fabb651e0a4979566fe83f0606fa92ebb28c/proto/stream_plan.proto#L654-L655
> 
> Great! is it also work for `create sink` statement?

Yes, it works for all streaming jobs.

**Comment by hzxa21** [comment]
> LGTM, and maybe we can store definition of the query in the tableFragment to get more information? c.c. @BugenZhao @yezizp2012

Sounds good to me. My only concern is to store mv definition in prometheus as a label since it is a relative long string. Not sure  whether prometheus will perform label deduplication in its storage.

**Pull request merged**